docs_src/design/adr/security/0009-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,207 @@
+
+- Secrets embedded in binaries
+
+  Binaries are usually not protected against confidentiality threats,
+  and binaries can be easily reverse-engineered to find any secrets therein.
+  Binaries included compile executables as well as Docker images.
+
+#### Recommended practices
+
+1. Direct consumption from process-to-process interaction with secret store
+
+   This approach is only possible for components that have native support for
+   [Hashicorp Vault](https://www.vaultproject.io/).
+   This includes any EdgeX service that links to go-mod-secrets.
+
+2. Dynamic injection of secret into process environment space
+
+   Environment variables are part of a process' environment block
+   and are mapped into a process' memory.
+   In this scenario,
+   an intermediary makes a connection to the secret store to fetch a secret,
+   store it into an environment variable,
+   and then launches a target executable,
+   thereby passing the secret _in-memory_ to the target process.
+
+   There are a few potential risks with this approach:
+
+   * Environment blocks are passed to child processes by default.
+   * Environment-variable-sniffing malware (introduced by compromised 3rd party libaries) is a proven attack method.
+
+3. Distribution via a secrets volume
+
+   This option is the most widely supported secret distribution mechanism by container orchestrators.
+   Docker Swarm secrets are shared via the `/run/secrets` volume,
+   which is a Linux `tmpfs` volume created on the host and shared with the container.
+   Kubernetes supports the same approach,
+   though the secrets volume can be mounted anywhere in the container namespace.
+
+   Only Linux platforms support `tmpfs` volumes for secret sharing,
+   and generally `/run` on a Linux system is a `tmpfs` volume.
+
+   * Docker:
+   For reasons of portability, `/tmp/edgex/secrets` has been chosen as a
+   cross-platform alternative to `/run`:
+   on the MacOS Docker implementation,
+   `/tmp` is a white-listed shared host directory,
+   while `/run` is not.
+   For production hardening, `/tmp` should be replaced by `/run`.
+
+   * Snaps:
+   Note that due to an existing [bug](https://bugs.launchpad.net/snap-confine/+bug/1620442),
+   `$XDG_RUNTIME_DIR` is set to a snap-specific path
+   (`/run/user/1000/snap.`_$SNAP_NAME_`/`),
+   but the directory is **neither created nor accessible to the snap**.
+   Because it may be necessary to share several secrets outside of the snap
+   --such as secret store tokens--
+   it is desirable to place them in a location that can be mapped to
+   a snap content interface, such as `$SNAP_DATA/secrets`.
+   Although not a `tmpfs`, this is the best available option at this time.
+   The snap bug should be escalated in priority.
+   
+4. Injection via command-line argument
+
+   This option is available as a **last resort**,
+   as process command-lines are _public information_ on most operating systems.
```

Comment:
BN - really want to remove this as one of the permitted options.

Reply:
Removed this option.  Instead, added "injection into configuration file" as an option between "injection via environment" and "secrets volume".